### ISSUE TITLE
Tensorflow: support int8 allreduce

### ISSUE THREAD

**Issue by kvignesh1420** [horovod/horovod#3642]
**Is your feature request related to a problem? Please describe.**
Currently, `HorovodAllreduce` tensorflow op is registered to support allreduce on `int32, int64, float16, float32, float64` types but not on `int8` even though the backends support it (ex: NCCL, MPI). 

**Describe the solution you'd like**
I plan to add `int8` to the supported types of `HorovodAllreduce`.

**Describe alternatives you've considered**
An alternative approach would be to use `allgather` on `int8` tensors and then do the addition/averaging in the user code, which is not efficient.

**Additional context**
With this in place, I believe we can also add an `Int8Compressor` which is extreme but people have used `1bit` compressors in the literature so `8bit` compression can definitely help.

**Comment by maxhgerlach** [comment]
Is there a specific reason HorovodAllreduce only supports `int32, int64, float16, float32` (other than avoiding integer overflows)? `HorovodAllgather` and `HorovodBroadcast` have supported `int8` and `uint8` since the initial commit.

**Comment by kvignesh1420** [comment]
I am not familiar with the initial design decisions but to address your question of integer overflows, I was thinking that in the compressor, we can:
1.  Threshold the values to be in the range `[-128, 127]`
2. Cast the thresholded tensor to int8
3. Cast it back to fp32 during decompression.

There will be loss of information with thresholding but at-least it would be deterministic.

**Comment by maxhgerlach** [comment]
In combination with a compressor `int8` could make a lot of sense!

I was just wondering if anybody recalled any TF kernel incompatibilities. For instance Horovod's in-place broadcast op for TF does not support `(u)int16` because there were no `DenseUpdate` kernels for those data types.

**Comment by romerojosh** [comment]
I don't recall there being any incompatibilities. Probably just an oversight early on in Horovod development.

**Comment by kvignesh1420** [comment]
I see. So shall I go ahead and enable int8 allreduce and add the Int8Compressor? WDYT @maxhgerlach @romerojosh

**Comment by romerojosh** [comment]
Yes, I think that is a great idea. I'd suggest splitting this into an initial PR that adds (u)int8 support to TF (and the other frameworks if necessary) with extended testing, then a PR for the compressor if that sounds good to you.

**Comment by kvignesh1420** [comment]
sounds good!

**Comment by maxhgerlach** [comment]
#3649 for TensorFlow is merged. Do you also want to look into PyTorch and MXNet?

**Comment by kvignesh1420** [comment]
@maxhgerlach yes, will take a look!
